Add zip support for `-file`

FtZPetruska opened the pull request:
Fixes #133

---

This PR adds zip support using [libzip](https://libzip.org/).

The logic used is as follow:
- When processing the files passed to `-file`, consider files ending with `.zip`
- Use `I_FindZip` (forwarded to `I_FindFile`) to ensure the file exists.
- Get a system-specific temporary directory
- Unpack the archive to that directory
- Pass the directory to `AutoLoadWADs` and `AutoLoadPatches` to discover WADs and dehacked files.

---

The implementation consists in two modules:
- tempdir, to get OS-specific temporary directories.
- zipfile, to unpack zip archives to a directory.

The `tempdir` module will use `mkdtemp` on *nix systems that provide it, `GetTempPath(2)` on Windows, or fallback to the current directory. You may notice the large block of checks for `mkdtemp` in the CMakeLists, this is mainly due to glibc having different needs to make the function available, and macOS having it in a completely different header.

I preferred adding this modules over using `I_GetTempDir` as the latter makes the assumption that all non-Windows platform use `/tmp` which is not necessarily correct.

The `zipfile` module uses `libzip` to access and decompress the zip file to the given directory. I have added all the necessary stuff for CMake to find and consume libzip and added it to the CI.

There is an odd packaging issue on Debian/Ubuntu where the CMake config shipped with `libzip-dev` contains executable targets that are not automatically installed, hence why `zipcmp`, `zipmerge`, and `ziptool` are also manually installed.

The vcpkg build of libzip has no encryption support since you presumably would not want to share your WADs in password-protected archives. But it does build additional support for the gzip and lzma algorithm. I did not enable zstd support due to issues with consuming the libzip in CMake.

There were several alternatives I considered:
- miniz, used by Woof, provides Zlib's inflate/deflate, zip read/write, and PNG writing. While it has no dependency, it requires either its source to be committed to this repo or using vcpkg.
- minizip(-ng), seems to be on-par feature-wise with libzip, the main caveat with this one is that there are two "version", the original `minizip` and its continuation `minizip-ng`, but distros like Debian are still stuck on the old `minizip` version.

---

As for testing, I tried to following use cases:
- pass an inexistant zip file -> ignored
- pass an unrelated zip file -> decompressed, but nothing loaded
- pass a valid zip file in the current directory with `-file ./foo.zip` -> successfully loaded
- pass a valid zip file in the current directory with `-file foo.zip` -> successfully loaded
- pass a valid zip file in `DOOMWADDIR` with `-file foo.zip` -> successfully loaded
- pass a valid zip file in parent directory with `-file ../foo.zip` -> successfully loaded

FtZPetruska committed:
Add module to get temporary directories.

This module uses either POSIX's mkdtemp, Win32's GetTempPath(2), or
returns the current directory.

FtZPetruska committed:
Add optional libzip dependency for zip support.

zipcmp, zipmerge, and ziptool are installed on Linux as the CMake
config depends on them being available.

FtZPetruska committed:
Add module to unzip file to a directory.

This module uses libzip to decompress zipfiles to a directory.

FtZPetruska committed:
d_main: consider zip files.

FtZPetruska committed:
Add I_FindZip to leverage common search paths.

rfomin commented:
> I preferred adding this modules over using `I_GetTempDir` as the latter makes the assumption that all non-Windows platform use `/tmp` which is not necessarily correct.

Just curious, what OS besides Windows doesn't have `/tmp`? `I_GetTempDir` has been used by Choco/Crispy and PrBoom+ for many years. I think such an OS will require some porting work anyway.

FtZPetruska commented:
> Just curious, what OS besides Windows doesn't have `/tmp`? `I_GetTempDir` has been used by Choco/Crispy and PrBoom+ for many years. I think such an OS will require some porting work anyway.

It's not so much that `/tmp` does not exists, but canonically, you're supposed to look up the value of `TMPDIR`. For example on my machine `TMPDIR` is some directory under `/var`. Plus in the implementation, if no environment is found it will default to `/tmp`.

rfomin commented on prboom2/src/dsda/tempdir.c:
```diff
@@ -0,0 +1,185 @@
+#include <string.h>
+#include <strsafe.h>
+#if defined(HAVE_GET_TEMP_PATH_2)
+#define winapi_GetTempDir GetTempPath2
```
We don't define `UNICODE`, so we must use the `GetTempPath2W`/`GetTempPathW` function to support non-Latin paths. Typically `%TEMP%` is found in the user's directory on Windows, with an often non-Latin name. I recommend just using `tempdir = M_getenv("TEMP");` it works for us in Woof/Crispy. See: https://github.com/chocolate-doom/chocolate-doom/blob/981da81f2adabd6eb82f1c3fc4ea27baecfccd82/src/m_misc.c#L572

FtZPetruska replied:
Heh, you're probably right. I'll just go ahead and use the existing constructs instead of making a whole new module.

FtZPetruska committed:
doomtype.h: Add PATHNAME_SEPARATOR constant.

FtZPetruska committed:
d_main: replace tempdir module with M_GetTempDir.

FtZPetruska committed:
Revert "Add module to get temporary directories."

This reverts commit b09e7a5e55c11de52c0dd352e43172ff23b00fe6.

kraflab commented on prboom2/src/SDL/i_system.c:
```diff
@@ -513,3 +513,8 @@ char* I_FindDeh(const char* wfname)
   return I_FindFile(wfname, ".deh");
 }
+
+char* I_FindZip(const char* wfname)
```
Please add an `I_RequireZip` as well so we have consistent interface options.

FtZPetruska replied:
Added in [640f573](https://github.com/kraflab/dsda-doom/pull/292/commits/640f573eb115a4b37863151e1b45dfb7cec4d62f).

kraflab commented on prboom2/src/d_main.c:
```diff
@@ -1510,6 +1512,29 @@ static void EvaluateDoomVerStr(void)
+static void D_AddZip(const char* zipped_file_name)
+{
+  dsda_string_t temporary_directory = { NULL, 0 };
+  char* full_zip_path = I_FindZip(zipped_file_name);
```
Please use the aforementioned `I_RequireZip` here. It's dsda-doom policy to abort if a requested file doesn't exist (and then you won't need the null check).

FtZPetruska replied:
Done in [607c428](https://github.com/kraflab/dsda-doom/pull/292/commits/607c42895cd79d3bb4f9dc66b98c79a357c0e835).

kraflab commented on prboom2/src/d_main.c:
```diff
@@ -1510,6 +1512,29 @@ static void EvaluateDoomVerStr(void)
 
+static void D_AddZip(const char* zipped_file_name)
+{
+  dsda_string_t temporary_directory = { NULL, 0 };
```
General comment for the whole PR. I generally prefer code to be written like this:
```c
char* full_zip_path;

full_zip_path = I_FindZip(zipped_file_name);
```
I.e., split the list of variables from any code. The code base isn't consistent on this point but I try to do this for new code.

kraflab replied:
Also, please leave out the `= { NULL, 0 }` here and leave the reliance on the init function. I'm not sure what I will do with this dsda_string_t in the future, but the goal is to not have any fragile code related to it (e.g., this constant struct might end up being wrong in the future).

FtZPetruska replied:
Got it 👍 

I made the style changes in [9fb2b1f](https://github.com/kraflab/dsda-doom/pull/292/commits/9fb2b1f4ac196728626ebe9b0780f85d67da66bb) and fixed the `{ NULL, 0 }` issue in [607c428](https://github.com/kraflab/dsda-doom/pull/292/commits/607c42895cd79d3bb4f9dc66b98c79a357c0e835).

kraflab commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,135 @@
+#include "../lprintf.h"
+
+#ifdef HAVE_LIBZIP
+#include "../m_file.h"
```
Please organize headers and paths like so for consistency:
```c
#include <stdio.h>
#include <zip.h>

#include "m_file.h"
#include "z_zone.h"

#include "dsda/utility.h"
```

FtZPetruska replied:
Done in [9fb2b1f](https://github.com/kraflab/dsda-doom/pull/292/commits/9fb2b1f4ac196728626ebe9b0780f85d67da66bb).

kraflab commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,135 @@
+
+#include "../lprintf.h"
+
+#ifdef HAVE_LIBZIP
```
Unless there is a very good reason, I would prefer to not add _optional_ dependencies but rather require them. I want people to have a consistent experience when using the port and something like loading zip files I would see as a basic element of the port. You can try to convince me otherwise though if you feel strongly about making it optional.

FtZPetruska replied:
Oh I have nothing against making it a requirement, the package is widely available anyway 😁

Changed in [285a7eb](https://github.com/kraflab/dsda-doom/pull/292/commits/285a7ebc8638a6ff5fce2442fc536988134ef11f).

FtZPetruska committed:
Make libzip a requirement.

FtZPetruska committed:
i_system: Add I_RequireZip.

FtZPetruska committed:
d_main: Use RequireZip.

FtZPetruska committed:
zipfile: Apply suggested style change.

- Sort headers.
- Split declaration and assignment when possible.

kraflab commented on prboom2/src/d_main.c:
```diff
@@ -1510,6 +1512,27 @@ static void EvaluateDoomVerStr(void)
+
+  dsda_UnzipFile(full_zip_path, temporary_directory.string);
+
+  AutoLoadWADs(temporary_directory.string);
```
When someone records a demo, it adds wad and deh files used to the footer of the demo file so that the launcher can autodetect the needed files to play back the demo and load them without prompting the user to choose a file.

It does not add autoload files (because these tend to be things NOT required to play back a demo, like cosmetic changes).

But the files inside a zip need to be treated as "real files" so that they 1) get added to the footer of recordings and 2) trigger loading of autoload files for the wads within.

You'll need to do two things to sort this out:

1) The AutoLoadWADs function needs to take an argument that chooses the source it assigns to files (currently it's `source_auto_load`, but for this use case we need `source_pwad`). Most likely this will need a rename like `LoadWADsAtPath`.

2) Instead of auto loading deh patches, we need to add them to the `-deh` command so they are registered as "real files" as well. So instead of `ProcessDehFile` in `AutoLoadPatches` we would need to do `dsda_AppendStringArg(dsda_arg_deh, filename)` (this is the same thing that happens when it detects a deh file in the `-file` command). I guess an option is to add a flag to `AutoLoadPatches` for whether the dehs should be processed immediately or not, and then give it a corresponding name like `LoadDehackedFilesAtPath`.

FtZPetruska replied:
Thank you for the suggestions, this seems to work as expected (record a demo by passing only a zip file and dsda automatically adding the pwads/deh patches when playing it back).

1. Changed in [ddc6b4b](https://github.com/kraflab/dsda-doom/pull/292/commits/ddc6b4bd24dd62198c1ee8fe896fdc3674d4e510). Renamed to `LoadWADsAtPath` as you suggested. Made the function take the `wad_source_t` as an argument. Replaced all the existing calls with `source_auto_load` and made the one in `D_AddZip` use `source_pwad`.
2. Changed in [8e7b5a9](https://github.com/kraflab/dsda-doom/pull/292/commits/8e7b5a94547d8f777b76ff76b74619ab86cf6629). Renamed to `LoadDehackedFilesAtPath` as you suggested. Similar idea as above, I added a bool argument indicating whether loading should be deferred or processed immediately.

kraflab commented on prboom2/src/doomtype.h:
```diff
@@ -116,6 +116,12 @@ typedef unsigned char byte;
 #endif
 
+#ifdef WIN32
+#define PATHNAME_SEPARATOR '\\'
```
Does this actually matter? I ask because there's a lot of code in this code base that builds paths with `/` and it seems to work fine 😄 

But maybe it doesn't work on every system, I don't know.

bkoropoff replied:
DOS and Windows have always accepted `/` as a separator at the API level, but individual programs can choke on them due to using `/` to start command line switches, etc.  It's probably fine if it goes straight into a Win32 API call, but you could run into trouble if you pass it to a 3rd-party library which does something unexpected like pass it to an external program as a parameter.

FtZPetruska replied:
Seems like libzip does not make use of external programs, so I'll just revert that change. Thanks for the insight 👍 

---

EDIT: Reverted in [d295b81](https://github.com/kraflab/dsda-doom/pull/292/commits/d295b81125fef9ce0969213e94f04fe49cb65c59).

FtZPetruska committed:
Revert "doomtype.h: Add PATHNAME_SEPARATOR constant."

This reverts commit 559562c0dcbae62c58f21cfde2b4d2750cdc3efb and
replaces its uses in 665e9794badb56d212672514cf460d0707b40534.

FtZPetruska committed:
d_main: Pass the source type to AutoLoadWADs.

This allows reusing the function for AutoLoad and zip archives.

Also rename the function to LoadWADsAtPath.

FtZPetruska committed:
d_main: Allow defering the loading of deh files.

This allows using the function for loading dehacked patches from a zip
and have them added to demo recordings.

Also rename the function to LoadDehackedFilesAtPath.

kraflab commented on prboom2/src/dsda/zipfile.h:
```diff
@@ -0,0 +1,29 @@
+//	DSDA zipfile support using libzip
+//
+
+#ifndef DSDA_ZIPFILE
```
Please use `__DSDA_ZIPFILE__` for consistency with the other header files. I'm aware that using underscores at the front of these macros is "wrong" but for now I'd like to keep the standard.

FtZPetruska replied:
My editor sort of freaks out with the leading `__`, but I can live with it 😛

kraflab commented on prboom2/src/dsda/zipfile.h:
```diff
@@ -0,0 +1,29 @@
+#define DSDA_ZIPFILE
+
+/**
+ * Unzips the file into the destination
```
You can remove this block, it's not that kind of code base :^)

kraflab commented on prboom2/src/dsda/zipfile.h:
```diff
@@ -0,0 +1,29 @@
+//
+// Copyright(C) 2022 by Pierre Wendling
```
2023 :^)

kraflab commented on prboom2/src/d_main.c:
```diff
@@ -1510,6 +1519,24 @@ static void EvaluateDoomVerStr(void)
+  full_zip_path = I_RequireZip(zipped_file_name);
+  dsda_InitString(&temporary_directory, I_GetTempDir());
+  dsda_StringCatF(&temporary_directory, "/%s/", dsda_BaseName(zipped_file_name));
+  M_MakeDir(temporary_directory.string, false);
```
The second argument of `M_MakeDir` says whether or not this code _requires_ the directory to be successfully created, so we had better send `true`. Otherwise this will return the error without handling it.

kraflab commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,134 @@
+    const char *file_name = zip_get_name(archive, i, ZIP_FL_UNCHANGED);
+
+    /* Intermediate directories show up as file */
+    if (dsda_HasFileExt(file_name, "/")) {
```
Do zips always use `/` when read using this library or might it be `\` as well? There is a function `dsda_BaseName` that returns the file name without leading directories and accounts for both slashes. That would return a pointer to the null terminator of the file name if it ends in a slash.

FtZPetruska replied:
Yes, after testing with various archives made on Windows and macOS, I observed that:
- The path separator is always `/`
- The name of a directory always ends with a `/`

kraflab commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,134 @@
+    zip_file_t *zipped_file;
+    zip_stat_t stat;
+    FILE *dest_file_handle;
+    const char *file_name = zip_get_name(archive, i, ZIP_FL_UNCHANGED);
```
Probably it will save some room if we calculate the path and then pass it to functions below, rather than calculating it in each function:
```c
dsda_string_t full_path;

dsda_StringPrintF(&full_path, "%s%s", destination_directory, file_name);

...

dsda_CreateSubDir(full_path);

...

dsda_CreateFile(full_path);

...

dsda_FreeString(&full_path);
```

kraflab commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,134 @@
+  return file_handle;
+}
+
+static void dsda_WriteContentToFile(zip_file_t *input_file, FILE *dest_file, zip_uint64_t data_size) {
```
Take a look at `dboolean M_WriteFile(char const *name, const void *source, size_t length)` in `m_file`. You can allocate the buffer, call `zip_fread`, and then call `M_WriteFile` using the path, rather than creating a file separately. `M_WriteFile` also manages removing the file if there's an error writing it. You should probably also check if the return value matches the expected length and raise an error if not.

FtZPetruska replied:
I went ahead and used a fixed-sized buffer instead as suggested by @bkoropoff.

kraflab commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,134 @@
+    zip_stat_index(archive, i, ZIP_FL_UNCHANGED, &stat);
+    if ((stat.valid & ZIP_STAT_SIZE) == 0) {
+      lprintf(LO_ERROR, "dsda_WriteZippedFilesToDest: Failed to read size of zipped file %s.\n", file_name);
+      return;
```
If anything goes wrong in this process, we can call `I_Error` for issues instead of using `lprintf`, since it's a requirement that files coming through `-file` do exist. The `I_Error` function shows an error popup on windows and prints to the terminal otherwise, and does not return. Maybe at some point there will be a case for zip files where it's acceptable that the process fails, and we would need to add an option to trigger or not trigger `I_Error`, but I don't think we need to manage that right now.

kraflab commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,134 @@
+  char *buffer;
+
+  buffer = (char *)Z_Malloc(data_size);
+  if (buffer == NULL) {
```
Zone memory functions handle errors themselves. It is impossible to unsuccessfully return from them.

FtZPetruska committed:
zipfile: Use correct date and style.

bkoropoff commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,134 @@
+static void dsda_WriteContentToFile(zip_file_t *input_file, FILE *dest_file, zip_uint64_t data_size) {
+  char *buffer;
+
+  buffer = (char *)Z_Malloc(data_size);
```
Typically you would use a fixed-size buffer and read/write in a loop to avoid loading the entire contents into memory at once, which could be a substantial amount of data for a wad with lots of texture replacements, ogg/mp3 music, etc.

bkoropoff commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,134 @@
+      zip_fclose(zipped_file);
+      return;
+    }
+    dsda_WriteContentToFile(zipped_file, dest_file_handle, stat.size);
```
You should refuse to load content over a certain cumulative size, e.g. more than 1GB of data from the same zip, to defend against [zip bombs](https://en.wikipedia.org/wiki/Zip_bomb).

bkoropoff commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,134 @@
+    lprintf(LO_ERROR, "dsda_WriteContentToFile: Could not allocate temporary buffer.\n");
+    return;
+  }
+  zip_fread(input_file, buffer, data_size);
```
Need to check for error (`-1`) here

bkoropoff commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,134 @@
+    return;
+  }
+  zip_fread(input_file, buffer, data_size);
+  fwrite(buffer, sizeof(char), data_size, dest_file);
```
Technically need to check for error here as well

FtZPetruska committed:
zipfile: Apply suggestions from review.

- Use a dsda string to manipulate paths.
- Use I_Error to print and exit on error.
- Read zip file by chunks to handle larger files.
- Guard against unzipping more than 1GB.

FtZPetruska committed:
d_main:  require M_MakeDir, remove trailing '/'.

All zip paths start with a pathname separator already.

FtZPetruska committed:
Revert "d_main: remove trailing '/'."

This partially reverts commit 0e9ca653dd69052d55b137adfcca52ce810972d1.

kraflab committed:
Adjust style in zipfile.c

kraflab commented on prboom2/src/dsda/zipfile.c:
```diff
@@ -0,0 +1,109 @@
+  }
+}
+
+void dsda_UnzipFile(const char *zipped_file_name, const char *destination_directory) {
```
We better set `total_bytes_read` to zero each time this function is called, right? In case we are loading multiple zip files.

FtZPetruska replied:
I wasn't sure whether to make `total_bytes_read` a global limit for all archives, or a per-archive limit. But I figured it makes more sense to reset it for every archive.

FtZPetruska committed:
zipfile: Reset total bytes count for each archive.

kraflab reviewed (approved):
Awesome stuff, thanks 👍
